File: Difficulty: Medium/Longest Sub-Array with Sum K/longest-sub-array-with-sum-k.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
class Solution{
    public:
    int lenOfLongSubarr(int A[],  int N, int K) 
    { 
        // Complete the function
        unordered_map<int,int> mpp; //<sum,idx>
        int sum=0;
        int maxLen=0;
        
        for(int i=0;i<N;i++){
            sum+=A[i];
            if(sum==K){ //if subarray starting from 0 idx.
                maxLen=max(maxLen,i+1);
            }
            if(mpp.find(sum-K)!=mpp.end()){
                maxLen=max(maxLen,i-mpp[sum-K]);
            }
            if(mpp.find(sum)==mpp.end()){
                mpp[sum]=i;
            }
        }
        return maxLen;
    } 

};

//{ Driver Code Starts.

int main() {
	//code
	
	int t;cin>>t;
	while(t--)
	{
	    int n, k;
	    cin>> n >> k;
	    int a[n];
	    
	    for(int i=0;i<n;i++)
	        cin>>a[i];
	   Solution ob;
	   cout << ob.lenOfLongSubarr(a, n , k)<< endl;
	    
	}
	
	return 0;
}
// } Driver Code Ends

File: Difficulty: Medium/Longest Subarray with Sum K/longest-subarray-with-sum-k.java
//{ Driver Code Starts
import java.io.*;
import java.util.*;

class Main {
    public static void main(String args[]) throws IOException {
        BufferedReader read = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(read.readLine().trim()); // Read number of test cases

        while (t-- > 0) {
            String line = read.readLine().trim(); // Read the array input
            String[] numsStr = line.split(" ");   // Split the input string by spaces
            int[] nums =
                new int[numsStr.length]; // Convert string array to integer array
            for (int i = 0; i < numsStr.length; i++) {
                nums[i] = Integer.parseInt(numsStr[i]);
            }

            int k = Integer.parseInt(read.readLine().trim()); // Read target sum

            Solution ob = new Solution();
            int ans = ob.longestSubarray(nums, k); // Call the function and store result
            System.out.println(ans);
            System.out.println("~");
        }
    }
}

// } Driver Code Ends


// User function Template for Java

class Solution {
    public int longestSubarray(int[] nums, int k) {
        // code here
        int n=nums.length;
        int maxLen=0;

        HashMap<Integer,Integer> mpp = new HashMap<>(); //map<prefixSum,idx>

        int prefixSum=0;
        for(int i=0;i<n;i++){
            prefixSum+=nums[i];
            
            if(prefixSum==k){
                maxLen = Math.max(i+1,maxLen);
            }
            
            if(mpp.containsKey(prefixSum-k)){
                maxLen = Math.max(i-mpp.get(prefixSum-k),maxLen);
            }
            
            if(!mpp.containsKey(prefixSum)){
                mpp.put(prefixSum,i);
            }
        }

        return maxLen;
    }
}
